completer.go

```diff
@@ -1,23 +1,43 @@
 package completer
 
-type Completer map[string]string
+import "fmt"
+
+type Completer struct {
+	aliases   map[string]string
+	originals map[string]struct{}
+}
 
 func NewCompleter() Completer {
-	return Completer(make(map[string]string))
+	return Completer{aliases: make(map[string]string), originals: make(map[string]struct{})}
 }
 
-func (c Completer) Add(s string) {
+func (c Completer) Add(s string) error {
+	a := c.aliases
+	if _, ok := a[s]; ok {
+		if _, ok := c.originals[s]; !ok {
+			return fmt.Errorf("unable to add %q: already present as an alias", s)
+		} else {
+			// Duplicate Add() - ignore
+			return nil
+		}
+	}
 	for i := 0; i < len(s); i++ {
 		prefix := s[:i+1]
-		if _, ok := c[prefix]; ok {
-			delete(c, prefix)
+		if _, ok := c.originals[prefix]; ok {
+			return fmt.Errorf("unable to add %q: substring %q already present", s, prefix)
```

Comment:
The code should be able to handle the case where added strings are prefixes of others, i.e.
```
c := NewCompleter()
c.Add("foo")
c.Add("fo") // no error
c.Lookup("foo") // returns "foo"
c.Lookup("fo") // returns "fo"
c.Lookup("f") // error, ambiguous
```

Reply:
Should it be an error or should it return `"", false`?

Reply:
I guess the current behavior is `"", false` so I'll keep that.